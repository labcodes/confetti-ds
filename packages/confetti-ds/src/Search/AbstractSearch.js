/* eslint-disable react/no-did-update-set-state */
import React from "react";
import PropTypes from "prop-types";
import { isUndefined } from "lodash";
import Icon from "../Icon";

export default class AbstractSearch extends React.Component {
  static propTypes = {
    /** Text that will serve as unique identifier. It's also an important accessibility tool. */
    id: PropTypes.string,
    /** Defines a default value for the Search initialization. */
    defaultValue: PropTypes.string,
    /** Disables the Search. Will be read by screen readers. When true, will override `disabled`. */
    ariaDisabled: PropTypes.bool,
    /** Disables the Search. Won't be read by screen readers. */
    disabled: PropTypes.bool,
    /** Text that will be rendered inside the Search field. */
    value: PropTypes.string,
    /** Action to be executed when the Search default value changes. */
    onChange: PropTypes.func,
    /** Action to be executed when the search is performed. */
    onSearch: PropTypes.func,
    /** Action to be executed when the Search field is cleared out. */
    onClear: PropTypes.func,
    /** The placeholder text when the Search field is empty. Usually used to describe the values accepted (e.g.: Search by keyword or status). */
    placeholder: PropTypes.string,
    /** Style variation of the Search. */
    type: PropTypes.oneOf(["standard", "inline"]).isRequired,
  };

  static defaultProps = {
    id: undefined,
    defaultValue: undefined,
    disabled: false,
    ariaDisabled: false,
    value: undefined,
    onChange: () => {},
    onSearch: () => {},
    onClear: () => {},
    placeholder: " ",
  };

  constructor(props) {
    super(props);
    this.searchRef = React.createRef();

    const { defaultValue, value, id } = props;
    if (!isUndefined(defaultValue) && !isUndefined(value)) {
      // eslint-disable-next-line no-console
      console.warn(
        `You are setting both value and defaultValue for input ${id} at the same time. When a value is passed, we always use it. Make sure this is the behaviour you want.`
      );
    }
    this.state = {
      localValue: value || defaultValue || "",
    };
  }

  componentDidUpdate(prevProps) {
    const { value } = this.props;

    if (value && value !== prevProps.value) {
      this.setState({ localValue: value });
    }
  }

  handleOnChange = (event) => {
    const { onChange } = this.props;

    this.setState({ localValue: event.target.value });

    if (!isUndefined(onChange)) {
      onChange(event);
    }
  };

  handleOnSearch = () => {
    const { onSearch } = this.props;
    const { localValue } = this.state;

    if (!isUndefined(onSearch)) {
      onSearch(localValue);
    }
  };

  handleKeyPress = (event) => {
    const { onSearch } = this.props;
    if (event.keyCode === 13 && !isUndefined(onSearch)) {
      onSearch(event.target.value);
    }
  };

  handleOnClear = () => {
    const { value, onClear } = this.props;

    if (isUndefined(value)) {
      this.setState({ localValue: "" });
    }

    if (!isUndefined(onClear)) {
      onClear();
    }
  };

  render() {
    const { id, disabled, ariaDisabled, placeholder, type } = this.props;

    const { localValue } = this.state;

    return (
      <div
        className={
          type === "standard" ? "lab-standard-search" : "lab-inline-search"
        }
      >
        <div
          className={`lab-search__wrapper ${
            disabled ? "lab-search--disabled" : ""
          }`}
        >
          <input
            className="lab-search__field"
            type="search"
            autoComplete="off"
            id={id}
            value={localValue}
            ref={this.searchRef}
            onChange={!ariaDisabled ? this.handleOnChange : () => {}}
            onKeyDown={!ariaDisabled ? this.handleKeyPress : () => {}}
            disabled={(!ariaDisabled && disabled) || undefined}
            aria-disabled={ariaDisabled || undefined}
            {...(placeholder ? { placeholder } : "")}
          />

          <div className="lab-search__borders" />
          <TrailingIcon
            onClear={this.handleOnClear}
            disabled={(!ariaDisabled && disabled) || undefined}
            ariaDisabled={ariaDisabled || undefined}
          />
          {type === "standard" ? (
            <StandardSearchIcon
              handleOnSearch={this.handleOnSearch}
              disabled={(!ariaDisabled && disabled) || undefined}
              ariaDisabled={ariaDisabled || undefined}
            />
          ) : (
            <InlineSearchIcon disabled={ariaDisabled || disabled} />
          )}
        </div>
      </div>
    );
  }
}

// ----- Auxiliary components ----- //

function TrailingIcon(props) {
  const { onClear, disabled, ariaDisabled } = props;
  let className = "lab-search__remove-icon";
  if (!onClear) {
    className += " lab-input__icon--disabled"; // check this out
  }

  return (
    <button
      type="button"
      className={className}
      onClick={!ariaDisabled ? onClear : () => {}}
      disabled={(!ariaDisabled && disabled) || undefined}
      aria-disabled={ariaDisabled || undefined}
    >
      <Icon type="remove" color="mineral-40" />
    </button>
  );
}

TrailingIcon.propTypes = {
  onClear: PropTypes.func,
  disabled: PropTypes.bool,
  ariaDisabled: PropTypes.bool,
};

TrailingIcon.defaultProps = {
  onClear: undefined,
  disabled: false,
  ariaDisabled: false,
};

function StandardSearchIcon(props) {
  const { disabled, ariaDisabled, handleOnSearch } = props;
  return (
    <React.Fragment>
      <button
        type="button"
        className="lab-standard-search__button"
        onClick={!ariaDisabled ? handleOnSearch : () => {}}
        disabled={(!ariaDisabled && disabled) || undefined}
        aria-disabled={ariaDisabled || undefined}
      >
        <Icon
          className="lab-standard-search__icon"
          type="magnifying-glass"
          color="white"
        />
      </button>
      <span className="lab-standard-search__separator" />
    </React.Fragment>
  );
}

StandardSearchIcon.propTypes = {
  disabled: PropTypes.bool,
  ariaDisabled: PropTypes.bool,
  handleOnSearch: PropTypes.func,
};

StandardSearchIcon.defaultProps = {
  disabled: false,
  ariaDisabled: false,
  handleOnSearch: undefined,
};

function InlineSearchIcon(props) {
  const { disabled } = props;
  return (
    <Icon
      className={`lab-inline-search__icon ${
        disabled ? "lab-inline-search__icon--disabled" : ""
      }`}
      type="magnifying-glass"
      color="mineral-40"
    />
  );
}

InlineSearchIcon.propTypes = {
  disabled: PropTypes.bool,
};

InlineSearchIcon.defaultProps = {
  disabled: false,
};
