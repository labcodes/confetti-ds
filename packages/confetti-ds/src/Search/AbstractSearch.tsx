import React, { useRef, useState, useEffect, SyntheticEvent } from "react";
import { isUndefined } from "lodash";
import Icon from "../Icon";
import { BaseSearchProps } from "./types";

interface AbstractSearchProps extends BaseSearchProps {
  /** Defines a default value for the Search initialization. */
  defaultValue?: string;
  /** Disables the Search. Will be read by screen readers. When true, will override `disabled`. */
  ariaDisabled?: boolean;
  /** Disables the Search. Won't be read by screen readers. */
  disabled?: boolean;
  /** The placeholder text when the Search field is empty. Usually used to describe the values accepted (e.g.: Search by keyword or status). */
  placeholder?: string;
  /** Defines which type of search will be renderized*/
  type?: "inline" | "standard";
  /** Previous valued stored on cache*/
  prevValue?: string;
}

export default function AbstractSearch({
  id,
  defaultValue,
  value,
  type,
  disabled = false,
  ariaDisabled = false,
  onChange = () => {},
  onSearch = () => {},
  onClear = () => {},
  placeholder = " ",
}: AbstractSearchProps) {
  const [localValue, setLocalValue] = useState(value || defaultValue || "");
  const searchRef = useRef<HTMLInputElement>();
  const prevValue = useRef<string>();

  const handleOnChange = (event) => {
    setLocalValue(event.target.value);
    if (!isUndefined(onChange)) {
      onChange(event);
    }
  };

  const handleOnSearch = (localValue) => {
    if (!isUndefined(onSearch)) {
      onSearch(localValue);
    }
  };

  const handleKeyPress = (event) => {
    if (event.keyCode === 13 && !isUndefined(onSearch)) {
      onSearch(event.target.value);
    }
  };

  const handleOnClear = (event) => {
    if (isUndefined(value)) {
      setLocalValue("");
    }

    if (!isUndefined(onClear)) {
      onClear(event);
    }
  };

  useEffect(() => {
    if (value && value !== prevValue.current) {
      setLocalValue(value);
    }
  }, [defaultValue]);

  if (!isUndefined(defaultValue) && !isUndefined(value)) {
    // eslint-disable-next-line no-console
    console.warn(
      `You are setting both value and defaultValue for input ${id} at the same time. When a value is passed, we always use it. Make sure this is the behaviour you want.`
    );
  }

  return (
    <div
      className={
        type === "standard" ? "lab-standard-search" : "lab-inline-search"
      }
    >
      <div
        className={`lab-search__wrapper ${
          disabled ? "lab-search--disabled" : ""
        }`}
      >
        <input
          className="lab-search__field"
          type="search"
          autoComplete="off"
          id={id}
          value={localValue}
          ref={searchRef}
          onChange={!ariaDisabled ? handleOnChange : () => {}}
          onKeyDown={!ariaDisabled ? handleKeyPress : () => {}}
          disabled={(!ariaDisabled && disabled) || undefined}
          aria-disabled={ariaDisabled || undefined}
          {...(placeholder ? { placeholder } : "")}
        />

        <div className="lab-search__borders" />
        <TrailingIcon
          onClear={handleOnClear}
          disabled={(!ariaDisabled && disabled) || undefined}
          ariaDisabled={ariaDisabled || undefined}
        />
        {type === "standard" ? (
          <StandardSearchIcon
            handleOnSearch={handleOnSearch}
            disabled={(!ariaDisabled && disabled) || undefined}
            ariaDisabled={ariaDisabled || undefined}
          />
        ) : (
          <InlineSearchIcon disabled={ariaDisabled || disabled} />
        )}
      </div>
    </div>
  );
}

// ----- Auxiliary components ----- //

interface TrailingIconProps {
  disabled?: boolean;
  ariaDisabled?: boolean;
  onClear?: (event?: SyntheticEvent) => any;
}

function TrailingIcon({
  disabled = false,
  ariaDisabled = false,
  onClear = () => {},
}: TrailingIconProps) {
  let className = "lab-search__remove-icon";
  if (!onClear) {
    className += " lab-input__icon--disabled"; // check this out
  }

  return (
    <button
      type="button"
      className={className}
      onClick={!ariaDisabled ? onClear : () => {}}
      disabled={(!ariaDisabled && disabled) || undefined}
      aria-disabled={ariaDisabled || undefined}
    >
      <Icon type="remove" color="mineral-40" />
    </button>
  );
}

interface StandardSearchIconProps {
  disabled?: boolean;
  ariaDisabled?: boolean;
  handleOnSearch?: (event?: SyntheticEvent) => any;
}
function StandardSearchIcon({
  disabled = false,
  ariaDisabled = false,
  handleOnSearch = () => {},
}: StandardSearchIconProps) {
  return (
    <React.Fragment>
      <button
        type="button"
        className="lab-standard-search__button"
        onClick={!ariaDisabled ? handleOnSearch : () => {}}
        disabled={(!ariaDisabled && disabled) || undefined}
        aria-disabled={ariaDisabled || undefined}
      >
        <Icon
          className="lab-standard-search__icon"
          type="magnifying-glass"
          color="white"
        />
      </button>
      <span className="lab-standard-search__separator" />
    </React.Fragment>
  );
}

function InlineSearchIcon({ disabled }: { disabled: boolean }) {
  return (
    <Icon
      className={`lab-inline-search__icon ${
        disabled ? "lab-inline-search__icon--disabled" : ""
      }`}
      type="magnifying-glass"
      color="mineral-40"
    />
  );
}
