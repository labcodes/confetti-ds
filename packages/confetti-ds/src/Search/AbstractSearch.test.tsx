/* eslint-disable no-console */
import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";

import AbstractSearch from "./AbstractSearch";

describe("AbstractSearch", () => {
  const originalWarn = console.warn;
  afterEach(() => {
    console.warn = originalWarn;
  });

  it("renders with base props", async () => {
    const renderedComponent = renderer
      .create(<AbstractSearch type="standard" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing disabled as true", async () => {
    const renderedComponent = mount(
      <AbstractSearch type="standard" disabled />
    );
    expect(renderedComponent.find("input[disabled]")).toHaveLength(1);
  });

  it("raises console.warn and sets state with value when passing value and defaultValue by props at the same time", async () => {
    console.warn = jest.fn();

    const component = mount(
      <AbstractSearch
        type="standard"
        defaultValue="default value"
        value="test value"
      />
    );

    expect(console.warn).toBeCalled();
    const inputElement = component.find("input");
    expect(inputElement.render().attr("value")).toBe("test value");
  });

  it("sets state with value if it is passed by props", async () => {
    const component = mount(
      <AbstractSearch type="standard" value="test value" />
    );

    const inputElement = component.find("input");
    expect(inputElement.render().attr("value")).toBe("test value");
  });

  it("sets state with defaultValue if it is passed by props and value is not passed by props", async () => {
    const component = mount(
      <AbstractSearch type="standard" defaultValue="default value" />
    );
    const inputElement = component.find("input");
    expect(inputElement.render().attr("value")).toBe("default value");
  });

  it("clears the input and calls onClear prop when the remove icon is clicked", async () => {
    const onClearFn = jest.fn();
    const component = mount(
      <AbstractSearch
        type="standard"
        defaultValue="default"
        onClear={onClearFn}
      />
    );

    expect(component.find("input").render().attr("value")).toBe("default");
    component.find(".lab-search__remove-icon").at(0).simulate("click");
    expect(component.find("input").render().attr("value")).toBe("");
    expect(onClearFn).toBeCalled();
  });

  it("calls onSearch prop when the magnifying glass icon is clicked and type is standard", async () => {
    const onSearchFn = jest.fn();
    const component = mount(
      <AbstractSearch
        type="standard"
        defaultValue="default"
        onSearch={onSearchFn}
      />
    );
    component.find(".lab-standard-search__button").at(0).simulate("click");
    expect(onSearchFn).toBeCalled();
  });

  it("calls onSearch prop when the enter key is pressed down", async () => {
    const onSearchStandard = jest.fn();
    const componentStandard = mount(
      <AbstractSearch
        type="standard"
        defaultValue="default"
        onSearch={onSearchStandard}
      />
    );
    componentStandard.find("input").at(0).simulate("keydown", { keyCode: 13 });
    expect(onSearchStandard).toBeCalled();

    const onSearchInline = jest.fn();
    const componentInline = mount(
      <AbstractSearch
        type="inline"
        defaultValue="default"
        onSearch={onSearchInline}
      />
    );
    componentInline.find("input").at(0).simulate("keydown", { keyCode: 13 });
    expect(onSearchInline).toBeCalled();
  });

  it("doesn't clear the input and calls onClear prop when ariaDisabled", async () => {
    const onClearFn = jest.fn();
    const component = mount(
      <AbstractSearch
        type="standard"
        ariaDisabled
        defaultValue="default"
        onClear={onClearFn}
      />
    );

    expect(component.find("input").render().attr("value")).toBe("default");
    component.find(".lab-search__remove-icon").at(0).simulate("click");
    expect(component.find("input").render().attr("value")).toBe("default");
    expect(onClearFn).not.toBeCalled();
  });

  it("doesn't call onSearch prop when ariaDisabled and clicked", async () => {
    const onSearchFn = jest.fn();
    const component = mount(
      <AbstractSearch
        ariaDisabled
        type="standard"
        defaultValue="default"
        onSearch={onSearchFn}
      />
    );
    component.find(".lab-standard-search__button").at(0).simulate("click");
    expect(onSearchFn).not.toBeCalled();
  });

  it("doesn't call onSearch prop when ariaDisabled and the enter key is pressed down", async () => {
    const onSearchStandard = jest.fn();
    const componentStandard = mount(
      <AbstractSearch
        ariaDisabled
        type="standard"
        defaultValue="default"
        onSearch={onSearchStandard}
      />
    );
    componentStandard.find("input").at(0).simulate("keydown", { keyCode: 13 });
    expect(onSearchStandard).not.toBeCalled();
  });
});
