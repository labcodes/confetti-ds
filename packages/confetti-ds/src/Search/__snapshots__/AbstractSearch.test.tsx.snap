// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbstractSearch renders with base props 1`] = `
<div
  className="lab-standard-search"
>
  <div
    className="lab-search__wrapper "
  >
    <input
      autoComplete="off"
      className="lab-search__field"
      id="test-search"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder=" "
      type="search"
      value=""
    />
    <div
      className="lab-search__borders"
    />
    <button
      className="lab-search__remove-icon"
      onClick={[Function]}
      type="button"
    >
      <span
        className="lab-icon lab-icon--remove"
      >
        <svg
          aria-hidden={true}
          className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
          fill="currentColor"
          height={20}
          viewBox="0 0 20 20"
          width={20}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m4.09 4.22.06-.07a.5.5 0 0 1 .63-.06l.07.06L10 9.29l5.15-5.14a.5.5 0 0 1 .63-.06l.07.06c.18.17.2.44.06.63l-.06.07L10.71 10l5.14 5.15c.18.17.2.44.06.63l-.06.07a.5.5 0 0 1-.63.06l-.07-.06L10 10.71l-5.15 5.14a.5.5 0 0 1-.63.06l-.07-.06a.5.5 0 0 1-.06-.63l.06-.07L9.29 10 4.15 4.85a.5.5 0 0 1-.06-.63l.06-.07-.06.07Z"
            fill="currentColor"
          />
        </svg>
      </span>
    </button>
    <button
      className="lab-standard-search__button"
      onClick={[Function]}
      type="button"
    >
      <span
        className="lab-icon lab-icon--lupe lab-standard-search__icon"
      >
        <svg
          aria-hidden={true}
          className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
          fill="currentColor"
          height={20}
          viewBox="0 0 20 20"
          width={20}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.5 3a5.5 5.5 0 0 1 4.23 9.02l4.12 4.13a.5.5 0 0 1-.63.76l-.07-.06-4.13-4.12A5.5 5.5 0 1 1 8.5 3Zm0 1a4.5 4.5 0 1 0 0 9 4.5 4.5 0 0 0 0-9Z"
            fill="currentColor"
          />
        </svg>
      </span>
    </button>
    <span
      className="lab-standard-search__separator"
    />
  </div>
</div>
`;
